dlc-manager/src/sub_channel_manager.rs

```diff
@@ -692,80 +697,218 @@ where
 
     /// Start force closing the sub channel with given [`ChannelId`].
     pub fn force_close_sub_channel(&self, channel_id: &ChannelId) -> Result<(), Error> {
-        let (mut signed, state) = get_sub_channel_in_state!(
-            self.dlc_channel_manager,
-            *channel_id,
-            Signed,
-            None::<PublicKey>
+        let mut sub_channel = self
+            .dlc_channel_manager
+            .get_store()
+            .get_sub_channel(*channel_id)?
+            .ok_or(Error::InvalidParameters(format!(
+                "Unknown sub channel {:?}",
+                channel_id
+            )))?;
+
+        match sub_channel.state {
+            SubChannelState::Offered(_) => self.force_close_offered_channel(sub_channel)?,
+            SubChannelState::Accepted(ref a) => {
+                let commitment_transactions = a.commitment_transactions.clone();
+
+                self.force_close_with_saved_commitment(sub_channel, &commitment_transactions)?;
+            }
+            SubChannelState::Signed(_) | SubChannelState::Finalized(_) => {
+                self.force_close_signed_channel(sub_channel, None)?;
+            }
+            SubChannelState::Confirmed(ref c) => {
+                let commitment_transactions = c.commitment_transactions.clone();
+                self.force_close_with_saved_commitment(sub_channel, &commitment_transactions)?;
+            }
+            SubChannelState::CloseOffered(c) => {
+                sub_channel.state = SubChannelState::Signed(c.signed_subchannel);
+                self.force_close_signed_channel(sub_channel, None)?;
+            }
+            SubChannelState::CloseAccepted(c) => {
+                self.ln_channel_manager.with_channel_lock_no_check(
+                    channel_id,
+                    &sub_channel.counter_party,
+                    |channel_lock| {
+                        self.ln_channel_manager.set_funding_outpoint(
+                            channel_lock,
+                            &lightning::chain::transaction::OutPoint {
+                                txid: c.signed_subchannel.ln_glue_transaction.txid(),
+                                index: 0,
+                            },
+                            c.ln_rollback.channel_value_satoshis,
+                            c.ln_rollback.value_to_self_msat,
+                        );
+                        Ok(())
+                    },
+                )?;
+                sub_channel.state = SubChannelState::Signed(c.signed_subchannel);
```

Comment:
Mmmm I feel maybe I should give up this rollback. It's there mainly because I was lazy to handle too many states and wanted to reuse code, but I'm thinking it might not have been a good idea.

Reply:
Actually I realized this rollback will not be persisted if the force closing fail, so I think it's not a big deal at the end.